mono/utils/mono-state.c

```diff
@@ -20,12 +20,188 @@ extern GCStats mono_gc_stats;
 
 // For AOT mode
 #include <mono/mini/mini-runtime.h>
+#include <mono/utils/mono-threads-debug.h>
+#include <mono/utils/mono-merp.h>
 
 #ifdef TARGET_OSX
 #include <mach/mach.h>
 #include <mach/task_info.h>
 #endif
 
+#include <sys/param.h>
+#include <sys/sysctl.h>
+#include <fcntl.h>
+
+typedef struct {
+	const char *directory;
+	MonoSummaryStage level;
+} MonoSummaryTimeline;
+
+static const char *configured_timeline_dir;
+static MonoSummaryTimeline log;
+
+static void
+file_for_summary_stage (const char *directory, MonoSummaryStage stage, gchar *buff, size_t sizeof_buff)
+{
+	g_snprintf (buff, sizeof_buff, "%s%scrash_stage_%d", directory, G_DIR_SEPARATOR_S, stage);
+}
+
+gboolean
+mono_summarize_set_timeline_dir (const char *directory)
+{
+	if (directory) {
+		configured_timeline_dir = strdup (directory);
+		return g_ensure_directory_exists (directory);
+	} else {
+		configured_timeline_dir = NULL;
+		return TRUE;
+	}
+}
+
+gboolean
+mono_summarize_timeline_start (void)
+{
+	memset (&log, 0, sizeof (log));
+
+	if (configured_timeline_dir) {
+		log.level = MonoSummarySetup;
+		log.directory = configured_timeline_dir;
+		return TRUE;
+	} else {
+		return FALSE;
+	}
+}
+
+void
+mono_summarize_double_fault_log (void)
+{
+	char out_file [200];
+	file_for_summary_stage (log.directory, MonoSummaryDoubleFault, out_file, sizeof(out_file));
+	int handle = g_open (out_file, O_WRONLY | O_CREAT, S_IWUSR | S_IRUSR | S_IRGRP | S_IROTH);
+	close(handle);
+}
+
+void
+mono_summarize_timeline_phase_log (MonoSummaryStage next)
+{
+	if (log.level == MonoSummaryNone)
+		return;
+
+	if (!log.directory)
+		return;
+
+	MonoSummaryStage out_level;
+	switch (log.level) {
+		case MonoSummarySetup:
+			out_level = MonoSummarySuspendHandshake;
+			break;
+		case MonoSummarySuspendHandshake:
+			out_level = MonoSummaryDumpTraversal;
+			break;
+		case MonoSummaryDumpTraversal:
+			out_level = MonoSummaryStateWriter;
+			break;
+		case MonoSummaryStateWriter:
+#ifdef TARGET_OSX
+			if (mono_merp_enabled ()) {
+				out_level = MonoSummaryMerpWriter;
+			} else
+#endif
+			{
+				out_level = MonoSummaryCleanup;
+			}
+			break;
+		case MonoSummaryMerpWriter:
+			out_level = MonoSummaryMerpInvoke;
+			break;
+		case MonoSummaryMerpInvoke:
+			out_level = MonoSummaryCleanup;
+			break;
+		case MonoSummaryCleanup:
+			out_level = MonoSummaryDone;
+			break;
+
+		case MonoSummaryDone:
+			MOSTLY_ASYNC_SAFE_PRINTF ("Trying to log crash reporter timeline, already at done %d\n", log.level);
+			return;
+		default:
+			MOSTLY_ASYNC_SAFE_PRINTF ("Trying to log crash reporter timeline, illegal state %d\n", log.level);
+			return;
+	}
+
+	// g_assert (out_level == next);
+
+	if (out_level != next)
```

Comment:
`g_assertf(out_level == next, "Log Error: Log transition to %d, actual expected next step is %d\n", next, out_level);`

Reply:
@luhenry in our discussion, you emphasized the importance of making sure that this writes to the files even when the expected and actual transitions don't match up. 

You are seeing the result of me removing an assertion that looks just like what you suggest above. It was previously working like this, when you last saw it. 

Making this change would make us not write to those files when the transition isn't expected. It would assert and double-fault.

Reply:
You misunderstood part of the point. I emphasized that logging to stderr is quite useless in the case here, since no one is ever going to check that log.

We have 2 possible cases:
 - it's not a problem for the transitions not to match, then logging here is not useful.
 - it's a problem for the transitions not to match, then we should double-fault and make sure we have a record of it.

What you explained me is it's very unlikely for the transitions not to be in the specified order, and the fact they wouldn't would be extremely weird and could possibly the indication of a bug, memory corruption or at least something worth terminating the process early for. This would clearly indicate to be a problem for the transitions not to match.